### PR TITLE
feat(zone-sdk): emit common_prefix on channel update

### PR DESCRIPTION
**Pull request by pradovic**
## 1. What does this PR implement?

LEZ and Logos SQL both need to recreate the non-finalized view on conflict, instead of applying only adopted/orphan diffs. To enable this this PR adds `common_prefix` with the helper function `canonical_chain()` enabling to recreate canonical view state by applying `common_prefix` + `adopted` (`canoical_chain()`) set.

## 2. Does the code have enough context to be clearly understood?

Yes

## 3. Who are the specification authors and who is accountable for this PR?

@davidrusu @pradovic 

## 4. Is the specification accurate and complete?

Yes

## 5. Does the implementation introduce changes in the specification?

No

## Checklist

> [!WARNING]  
> Do not merge the PR if any of the following is missing:

* [ ] 1. The PR title follows the Conventional Commits [specification](https://www.notion.so/How-to-open-a-Pull-Request-215261aa09df80deb538ed59f5e4e0b5).
* [ ] 2. Description added.
* [ ] 3. Context and links to Specification document(s) added.
* [ ] 4. Main contact(s) (developers and specification authors) added
* [ ] 5. Implementation and Specification are 100% in sync including changes. This is critical.
* [ ] 6. Link PR to a specific milestone.
* [ ] 7. New or changed logs were reviewed against the [logging guidelines](https://github.com/logos-blockchain/logos-blockchain/blob/master/CONTRIBUTING.md#logging-guidelines).

### CODE REVIEW

File: tests/src/cucumber/steps/zone/operations/custom_policy.rs [hunk 1]
```diff
@@ -129,21 +126,14 @@ where
                 .iter()
                 .flat_map(|entry| self.entry_payloads(entry))
                 .collect();
-            let adopted: Vec<Inscription> = channel_update
-                .adopted
-                .iter()
+            let view: HashSet<Inscription> = channel_update
```
Comment (danielSanchezQ) [inline tests/src/cucumber/steps/zone/operations/custom_policy.rs:129]: Just to double check: Is ordering important? Before we had a vec and ordering may be wrong from the set.

File: tests/src/cucumber/steps/zone/runner.rs [hunk 2]
```diff
@@ -85,14 +89,74 @@ where
...
+            }
+        }
+        for tx in &self.held {
+            if !view.contains(tx) {
+                fail(&format!(
+                    "{tx:?} was adopted and never orphaned or finalized, but left common_prefix"
+                ));
+            }
+        }
+    }
```
Comment (danielSanchezQ) [inline tests/src/cucumber/steps/zone/runner.rs]: It would be nice to read this function as a list of operations:

```rust
fn observe() {
  let Event:: ...

  ...

   remove_orphaned()
   add_adopted()
   remove_finalized()
   ...
}
```

**Review by davidrusu — APPROVED** [review]

**Review by davidrusu — APPROVED** [review]

**Review by danielSanchezQ — APPROVED** [review]

File: zone-sdk/src/sequencer/state.rs [hunk 3]
```diff
@@ -1421,6 +1421,21 @@ impl TxState {
...
+    #[must_use]
+    pub fn channel_view_txs(
+        &self,
+        tip: HeaderId,
+        finalized_now: &HashSet<MsgId>,
+    ) -> Vec<ChannelUpdateTx> {
+        let lineage = self.channel_lineage(tip);
+        let mut finalized = self.finalized_prefix_ids(&lineage);
+        finalized.extend(finalized_now.iter().copied());
+        finalized.insert(self.finalized_config);
```
Comment (andrussal) [inline zone-sdk/src/sequencer/state.rs]: Should we include earlier finalized configs here too, like the prefix above?

Comment (pradovic) [reply zone-sdk/src/sequencer/state.rs]: Will be fixed in the PR before and rebased.

**Review by andrussal — APPROVED** [review]

File: tests/src/cucumber/steps/zone/runner.rs (hunk 2 above)
Comment (pradovic) [reply tests/src/cucumber/steps/zone/runner.rs]: Nice, fixed!

File: tests/src/cucumber/steps/zone/operations/custom_policy.rs (hunk 1 above)
Comment (pradovic) [reply tests/src/cucumber/steps/zone/operations/custom_policy.rs:129]: I think not really, membership only. `view` is only queried with contains, same as the pending set it replaced; the old Vec was extended straight into that set.

File: zone-sdk/src/sequencer/block_fetch.rs [hunk 4]
```diff
@@ -333,9 +335,17 @@ fn apply_prepared_block_event(
         update
     });
 
+    let old_txs: HashSet<TxHash> = old_lineage.iter().map(|info| info.tx_hash).collect();
+    let common_prefix = s
+        .channel_view_txs(tip, &finalized_now)
+        .into_iter()
+        .filter(|tx| old_txs.contains(&tx.tx_hash()))
+        .collect();
```
Comment (hansieodendaal) [inline zone-sdk/src/sequencer/block_fetch.rs:347]: **Invariant:**  After resuming from a checkpoint, the first emitted `canonical_chain()` must still contain the complete non-finalized view, including restored pending transactions.

Could we add a regression test for the first block event after restoring a checkpoint with an existing pending transaction? With the current code it looks like the pending transaction can be filtered out of `adopted` as already tracked while `common_prefix` is still empty.

File: zone-sdk/src/sequencer/block_fetch.rs [hunk 5]
```diff
@@ -333,9 +335,17 @@ fn apply_prepared_block_event(
         update
     });
 
+    let old_txs: HashSet<TxHash> = old_lineage.iter().map(|info| info.tx_hash).collect();
+    let common_prefix = s
+        .channel_view_txs(tip, &finalized_now)
+        .into_iter()
+        .filter(|tx| old_txs.contains(&tx.tx_hash()))
```
Comment (hansieodendaal) [inline zone-sdk/src/sequencer/block_fetch.rs:346]: **Invariant:** `canonical_chain()` preserves the ordering of the complete non-finalized view.

Could we add a regression test where the message and config lineages move independently?

For example, if pending config `C` is already in the view and the next block adopts inscription `M`, `channel_view_txs()` can produce:

```text
M, C
```

but the current split seems to put `C` in `common_prefix` and `M` in `adopted`, making:

```text
canonical_chain() = C, M
```

**Review by hansieodendaal — COMMENTED** [review]
Nice improvement, this is much cleaner for consumers.

Some comments to consider please.

File: zone-sdk/src/sequencer/block_fetch.rs (hunk 4 above)
Comment (pradovic) [reply zone-sdk/src/sequencer/block_fetch.rs:347]: Nice, thanks! https://github.com/logos-blockchain/logos-blockchain/pull/3634/changes/f13138f7c6013bdeed22efc733a2047603d22964

File: zone-sdk/src/sequencer/block_fetch.rs (hunk 5 above)
Comment (pradovic) [reply zone-sdk/src/sequencer/block_fetch.rs:346]: Thanks! https://github.com/logos-blockchain/logos-blockchain/pull/3634/changes/f13138f7c6013bdeed22efc733a2047603d22964

**Review by hansieodendaal — APPROVED** [review]
LGTM!

File: zone-sdk/src/sequencer/block_fetch.rs [hunk 6]
```diff
@@ -335,11 +335,16 @@ fn apply_prepared_block_event(
...
+    // was on the old lineage or pending before this event.
     let old_txs: HashSet<TxHash> = old_lineage.iter().map(|info| info.tx_hash).collect();
     let common_prefix = s
         .channel_view_txs(tip, &finalized_now)
         .into_iter()
-        .filter(|tx| old_txs.contains(&tx.tx_hash()))
+        .filter(|tx| {
+            let hash = tx.tx_hash();
+            old_txs.contains(&hash) || tracked_before.contains(&hash)
+        })
```
Comment (hansieodendaal) [inline zone-sdk/src/sequencer/block_fetch.rs]: **Invariant:** After a checkpoint restore, `canonical_chain()` preserves message-lineage order as well as completeness.

Could we extend the restore regression test to a pending `P` whose parent is an unfinalized mined message `M` that is rediscovered on the first event?

It looks like the first block-processing event after restart can produce:

```text
view:          M, P
adopted:       M
common_prefix: P
```

because `P` is already in `tracked_before`, giving this incorrect lineage:

```text
canonical_chain() = P, M
```

**Review by hansieodendaal — COMMENTED** [review]
Missed this (new) possible edge case

File: tests/src/cucumber/steps/zone/runner.rs [hunk 7]
```diff
@@ -85,14 +89,91 @@ where
...
+}
+
+/// Log, then panic: the drive task's panic is swallowed by cucumber's panic
+/// hook, so the log line is what the scenario output shows.
+fn fail(message: &str) -> ! {
+    error!("channel view contract violated: {message}");
+    panic!("channel view contract violated: {message}");
+}
```
Comment (hansieodendaal) [inline tests/src/cucumber/steps/zone/runner.rs]: The background checker should not panic or fail here. It should record the first contract violation, and then an explicit cucumber step should assert that no violation was recorded. We need:
- no hidden background panic,
- deterministic cucumber failure reporting,
- the invariant to be visible in the Cucumber feature file.

The checker would conceptually do something like:
```rust
self.violation.get_or_insert(message);
```
and the scenario would contain an explicit step such as:
```
Then the channel update view contract holds for sequencer "SEQ_A"
```
That step would fail normally via `StepError` if a violation had been recorded.

**Review by hansieodendaal — COMMENTED** [review]
One more comment on our Cucumber assertions - it must fit the Cucumber paradigm.

File: tests/src/cucumber/steps/zone/runner.rs [hunk 8]
```diff
@@ -80,19 +90,111 @@ pub struct Runtime {
...
+        }
+    }
+
+    fn remove_finalized(&mut self, finalized: &[FinalizedTx]) {
+        for tx in finalized {
+            self.held.remove(&tx.tx_hash);
+        }
+    }
+
+    /// `common_prefix ++ adopted` carries only what is held or in flight, and
```
Comment (hansieodendaal) [inline tests/src/cucumber/steps/zone/runner.rs]: Could we also assert the ordering part of the `canonical_chain()` contract as well?

`canonical_chain()` is immediately converted to a `HashSet`, so this currently verifies completeness/membership but cannot distinguish:

```text
M, P
```

from:

```text
P, M
```

**Review by hansieodendaal — COMMENTED** [review]
Thanks for fixing the cucumber paradigm!

One follow-up comment on that below, and then I believe [the previous lineage comment](https://github.com/logos-blockchain/logos-blockchain/pull/3634#discussion_r4084642462) still needs to be addressed.

File: zone-sdk/src/sequencer/block_fetch.rs (hunk 6 above)
Comment (pradovic) [reply zone-sdk/src/sequencer/block_fetch.rs]: Good catch, that was real. 

The cause was the first event after a restore: the SDK had no old view, so it reported the restored pending's mined ancestors as adopted while the pending itself sat in common_prefix, giving [P, M]. 

Fixed at the source: restored pending proves what it chains on, so on the first event the ancestors of tracked entries (new TxState::lineage_under) are treated as the view before the restart. Your case now yields no update at all and common_prefix = [M, P]. 

canonical_chain() is in lineage order on every event, and the runner checker now asserts parent order on every event. Covered by restored_pending_survives_its_parent_being_rediscovered, plus a counterpart where a rival takes M's slot and P is orphaned.

File: tests/src/cucumber/steps/zone/runner.rs (hunk 8 above)
Comment (pradovic) [reply tests/src/cucumber/steps/zone/runner.rs]: Added to check_view: every inscription in canonical_chain() must come after its parent when the parent is in the chain. Records a violation like the other checks, applies to every event.
